src/solver/multimarginloss/forward_reduced_multimarginloss.cpp

```diff
@@ -0,0 +1,237 @@
+            auto w_tv             = get_inner_expanded_tv<1>(deref(params.wDesc));
+
+            float elapsed = 0.0f;
+            HipEventPtr start;
+            HipEventPtr stop;
+
+            bool reset_profiling_state = false;
+            if(handle_.IsProfilingEnabled())
+            {
+                reset_profiling_state = true;
+                handle_.EnableProfiling(false);
+                start = miopen::make_hip_event();
+                stop  = miopen::make_hip_event();
+                hipEventRecord(start.get(), handle_.GetStream());
+            }
+            /* Phase 1: Calc loss for each element. */
+            {
+                decltype(auto) kernel = handle_.Run(kernels.front());
+                kernel(params.i,
+                       params.t,
+                       params.w,
+                       params.workspace,
+                       params.p,
+                       params.margin,
+                       params.divisor,
+                       i_tv,
+                       t_tv,
+                       w_tv);
+            }
+
+            /* Phase 2: Reduce */
+            auto size      = deref(params.iDesc).GetLengths()[0];
+            auto reduce_in = params.workspace;
+            auto reduce_out =
+                static_cast<Data_t>(static_cast<char*>(params.workspace) +
+                                    size * get_data_size(deref(params.oDesc).GetType()));
+
+            for(int i = 1; i < kernels.size(); ++i)
+            {
+                decltype(auto) kernel = handle_.Run(kernels[i]);
+                if(i + 1 != kernels.size())
+                {
+                    kernel(reduce_in, reduce_out, size);
+                    std::swap(reduce_in, reduce_out);
+                }
+                else
+                {
+                    kernel(reduce_in, params.o, size);
+                }
+                size = AlignUp(size, LOCAL_SIZE_REDUCE) / LOCAL_SIZE_REDUCE;
+            }
+
+            if(reset_profiling_state)
+                handle_.EnableProfiling(true);
+            if(handle_.IsProfilingEnabled())
+            {
+                hipEventRecord(stop.get(), handle_.GetStream());
+                hipEventSynchronize(stop.get());
+                hipEventElapsedTime(&elapsed, start.get(), stop.get());
+
+                // Clean up
+                hipEventDestroy(start.get());
+                hipEventDestroy(stop.get());
+                handle_.ResetKernelTime();
+                handle_.AccumKernelTime(elapsed);
+            };
+        };
+    };
+
+    return result;
+}
+
+std::size_t MultiMarginLossForward::GetWorkspaceSize(
+    const ExecutionContext& /*context*/,
+    const miopen::multimarginloss::ForwardProblemDescription& problem) const
+{
+    auto elem = problem.GetiDesc().GetLengths()[0];
+    return (elem + (elem + LOCAL_SIZE_REDUCE) / LOCAL_SIZE_REDUCE) *
+           get_data_size(problem.GetoDesc().GetType());
```

Comment:
Probably it does not take proper alignment into account.
`MultiBufferWorkspaceTraits` is a good option to use in this case.